ssl/record/dtls1_bitmap.c

```diff
@@ -13,83 +13,26 @@
 /* mod 128 saturating subtract of two 64-bit values in big-endian order */
 static int satsub64be(const unsigned char *v1, const unsigned char *v2)
 {
-    int ret, i;
+    int64_t ret;
+    uint64_t l1, l2;
 
-    if (sizeof(long) == 8)
-        do {
-            const union {
-                long one;
-                char little;
-            } is_endian = {
-                1
-            };
-            long l;
+    n2l8(v1, l1);
+    n2l8(v2, l2);
 
-            if (is_endian.little)
-                break;
-            /* not reached on little-endians */
-            /*
-             * following test is redundant, because input is always aligned,
-             * but I take no chances...
-             */
-            if (((size_t)v1 | (size_t)v2) & 0x7)
-                break;
+    ret = l1 - l2;
 
-            l = *((long *)v1);
-            l -= *((long *)v2);
-            if (l > 128)
-                return 128;
-            else if (l < -128)
-                return -128;
-            else
-                return (int)l;
-        } while (0);
-
-    ret = 0;
-    for (i=0; i<7; i++) {
-        if (v1[i] > v2[i]) {
-            /* v1 is larger... but by how much? */
-            if (v1[i] != v2[i] + 1)
-                return 128;
-            while (++i <= 6) {
-                if (v1[i] != 0x00 || v2[i] != 0xff)
-                    return 128; /* too much */
-            }
-            /* We checked all the way to the penultimate byte,
-             * so despite higher bytes changing we actually
-             * know that it only changed from (e.g.)
-             *       ... (xx)  ff ff ff ??
-             * to   ... (xx+1) 00 00 00 ??
-             * so we add a 'bias' of 256 for the carry that
-             * happened, and will eventually return
-             * 256 + v1[7] - v2[7]. */
-            ret = 256;
-            break;
-        } else if (v2[i] > v1[i]) {
-            /* v2 is larger... but by how much? */
-            if (v2[i] != v1[i] + 1)
-                return -128;
-            while (++i <= 6) {
-                if (v2[i] != 0x00 || v1[i] != 0xff)
-                    return -128; /* too much */
-            }
-            /* Similar to the case above, we know it changed
-             * from    ... (xx)  00 00 00 ??
-             * to     ... (xx-1) ff ff ff ??
-             * so we add a 'bias' of -256 for the borrow,
-             * to return -256 + v1[7] - v2[7]. */
-            ret = -256;
-        }
-    }
-
-    ret += (int)v1[7] - (int)v2[7];
+    /* We do not permit wrap-around */
+    if (l1 > l2 && ret < 0)
```

Comment:
Yes and no. You are entirely correct that the naïve reader, on idly reading your version for entertainment, doesn't actually need to know how integer wraparound works. You carefully avoid it, and they don't *need* to know that a big enough unsigned number becomes negative on being cast to signed.

That reader won't understand *why* you wrote the code that way, though. That whole `-(int)l2` thing, and the entirely different code path for the `(l2 > l1)` vs. the `(l1 > l2)` case, are going to look entirely strange to them. And they're going to be tempted to simplify it. Because the reader we most need to care about isn't just *reading* the code; they're thinking about *modifying* it.

My version puts a big sign up above the landmine which exists here. Yours takes a meandering path through the minefield which carefully happens to avoid triggering it.

It is my experience, and my firm belief, that those who come after us will benefit more from my approach than yours :)